src/main/java/org/fsg1/fmms/backend/services/CurriculumService.java

```diff
@@ -64,13 +73,11 @@ private ObjectNode buildCurriculumSemesters(final String jsonString) throws IOEx
         ArrayNode arrayOfModules = (ArrayNode) mapper.readTree(jsonString);
         if (arrayOfModules.size() == 0) return resultObject;
 
-        Set<Integer> seenSemesters = new HashSet<>();
+        Map<Integer, ArrayNode> seenSemesters = new HashMap<>();
```

Comment:
Please initialize the HashMap with a proper capacity which should be at least (1 + numSemester/0,75)